Fix for resize events when building with ASSERTIONS

cwoffenden opened the pull request:
Building with `ASSERTIONS` is failing the safe heap checks for resize events given that `e.detail` is seen as `undefined` (the specs say `UIEvent.detail` [should be zero](https://developer.mozilla.org/en-US/docs/Web/API/UIEvent/detail)). Seen here with the latest Emscripten 3.1.54, tested in Safari and Chrome on Mac.

<img width="385" alt="detail" src="https://github.com/emscripten-core/emscripten/assets/8437014/f407d0ff-2723-4cdb-86b3-1b29c2fcc56f">

The resulting error is:
```
Uncaught RuntimeError: Aborted(Assertion failed: attempt to write non-integer (undefined) into integer heap)
```
It's reported specifically in #21091 which is closed in favour of #20992. This fixes #21091 which fails without SDL, just with a simple `emscripten_set_resize_callback()` and `-s ASSERTIONS`.

sbc100 commented on src/library_html5.js:
```diff
@@ -710,7 +710,7 @@ var LibraryHTML5 = {
 #endif
       // e.detail is seen as undefined
-      {{{ makeSetValue('uiEvent', C_STRUCTS.EmscriptenUiEvent.detail, '((typeof e.detail !== "undefined") ? e.detail : 0), 'i32') }}};
+      {{{ makeSetValue('uiEvent', C_STRUCTS.EmscriptenUiEvent.detail, '(typeof e.detail !== "undefined") ? e.detail : 0', 'i32') }}};
```
Can this just be `e.detail || 0`?

cwoffenden replied:
Yes, it has the same result in the end.

cwoffenden commented:
A simple test, which also shows issues from #20992 is to build the HTML5 test, which already tests the resize:

https://github.com/emscripten-core/emscripten/blob/7cdfa8d9f22562e7c08789a5ae29ad9a1025a8e4/test/test_html5_core.c#L261

Building with `-s ASSERTIONS=2` will assert when resizing (it also asserts for other events, which I can probably look into as I come across them).

sbc100 reviewed (commented):
Thanks for the PR!

Do you think we write a test for this?

I think we have some tests that simulate events.. perhaps we could simulate and event without any detail?

cwoffenden commented:
> I think we have some tests that simulate events.. perhaps we could simulate and event without any detail?

It's been a while since I've used Emscripten but I don't recall this being broken last year (we build with asserts for debug at work and it would have cropped up), or why this wouldn't be to spec.

`test_html5_core.c` does test everything with some manual intervention, showing other asserts which will need fixing. The first I see is for the orientation, which is still `-1` as it reaches here:
https://github.com/emscripten-core/emscripten/blob/7cdfa8d9f22562e7c08789a5ae29ad9a1025a8e4/src/library_html5.js#L942
Which then asserts for the `i32` test.

I guess the answer is to first see why `detail` isn't defined (which I don't mind delving into, just not today).

sbc100 commented:
The reason this didn't occur in older versions of emscripten is that these asserts were only added in #19251.    And that PR didn't get included in a release until 3.1.38.

sbc100 commented:
I wonder if we can / should simply remove this line?

It looks like the `detail` property is only set for `mousedown` and `click` UIEvents (https://developer.mozilla.org/en-US/docs/Web/API/UIEvent/detail) but this code is only used for `resize` and `scroll`.

sbc100 commented:
Indeed, it looks like we should maybe completely remove this line.  Or maybe just set it 0 in all cases.

Ideally we would even remove the field from the EmscriptenUiEvent struct, but that could break folks who are reading it for some reason.

Perhaps add a comment to the field in `html5.h` saying that this field is deprecated and will always be zero?

sbc100 commented:
(Can you also remove the documentation for this field from `site/source/docs/api_reference/html5.h.rst`?)

cwoffenden commented:
I'll take a proper look next week. The docs do say: _For all other UIEvent objects, UIEvent.detail is always zero_, so I guess setting the struct field to zero is the best option, ignoring the `detail` completely and updating the docs to say just this. It keeps backwards compatibility and passes assertions.

cwoffenden commented:
I updated the PR as per @sbc100's comments, notes as follows:

- I didn't remove `.detail` since it would probably cause breakage for people, but since the spec says zero I just set it to zero (no point even checking if the browser's event contains a `detail`).
- I updated the docs to reflect this.
- Scroll events in the same section of code are broken with `ASSERTIONS` since `pageXOffset` and `pageYOffset` are floats, and the `EmscriptenUiEvent`'s `scrollTop` and `scrollLeft` are ints. It would also probably cause breakage if these were changed from ints, so to keep the behaviour the same I truncated them (which is the same as `HEAP32[((ptr) >> 2)] = value` is doing when they're set).

The existing HTML5 test does cover both these cases.

If this an acceptable fix, I can take a look at #20992 in a week or so when I have from free time (it's the same, just requires testing).
